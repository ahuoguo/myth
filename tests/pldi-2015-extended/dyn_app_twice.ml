type nat =
   | O
   | S of nat

type dyn =
  | Error
  | Base of nat
  | Dyn of (dyn -> dyn)


let succ (d:dyn) : dyn =
  match d with
  | Error -> Error
  | Base ( n ) -> Base (S(n))
  | Dyn ( f ) -> Error
;;

let pred (d:dyn) : dyn =
  match d with
  | Error -> Error
  | Base ( n ) ->
    (match n with
    | O -> Base ( O )
    | S ( n ) -> Base ( n ))
  | Dyn ( f ) -> Error
;;

let dyn_app_twice : dyn -> dyn -> dyn |>
{
  Dyn (succ) => ( Base( 0 ) => Base( 2 )
                | Base( 1 ) => Base( 3 ) )
| Dyn (pred) => ( Base( 0 ) => Base( 0 )
                | Base( 1 ) => Base( 0 ) )
| Error => Error => Error
| Base( 0 ) => Error => Error
} = ?
