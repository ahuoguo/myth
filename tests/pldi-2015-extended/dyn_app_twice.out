(* Generated Thu Nov 13 01:55:42 2014 *)
let dyn_app_twice : dyn -> dyn -> dyn =
  fun (d1:dyn) ->
    fun (d2:dyn) ->
      match d1 with
        | Error -> Error
        | Base (n1) -> Error
        | Dyn (f3) -> f3 (f3 d2)
;;
